### PR TITLE
move transaction summary computation to separate crate

### PR DESCRIPTION
**Pull request by ryankurte**
the second part of refactoring TxSummary computation into a separate mc-transaction-summary crate to simplify compilation for hardware wallets, and another step in the upstreaming of [ryan/ledger-wip](https://github.com/mobilecoinfoundation/mobilecoin/compare/ryan/ledger-wip).

- adds `transaction/summary` crate
- moves summary computation from `transaction/core`
- removes previous `UnmaskedAmount` definition
- propagates `transaction/types` as required
- refactors streaming summariser to simplify embedded use (while maintaining the external interface)

blocked on: 
- [x] #3125

### CODE REVIEW

**Comment by briancorbin** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3083** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3083" target="_blank">Graphite</a> 
    * **PR #3125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3125" target="_blank">Graphite</a> 
      * **PR #3132** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3132" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3132?utm_source=stack-comment).

**Review by briancorbin — APPROVED** [review]

File: transaction/summary/src/verifier.rs [hunk 1]
```diff
@@ -117,7 +46,17 @@ pub fn verify_tx_summary(
 /// implementation details of the mc-crypto-digestible scheme.
 /// If TxSummary digestible annotations are changed then this object's
 /// implementation needs to change also.
-pub struct TxSummaryStreamingVerifier {
+pub struct TxSummaryStreamingVerifier<const RECORDS: usize = MAX_RECORDS> {
```
Comment (ryankurte) [inline transaction/summary/src/verifier.rs:49]: `TxSummaryStreamingVerifier` only ended up being used internally for `verify_tx_summary` (because separating the verifier and report helps with stack use), could make this non-public

File: transaction/summary/src/data.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+/// Exercise the functionality of the streaming verifier, and return its
+/// results.
+///
+/// This is mainly useful for testing / demonstration purposes, since the more
+/// interesting use-case is when the streaming verifier is on a small remote
+/// device and doesn't have the full TxSummary or TxSummaryUnblindingData on
+/// hand.
+pub fn verify_tx_summary(
```
Comment (ryankurte) [inline transaction/summary/src/data.rs:75]: since `verify_tx_summary` uses the verifier over types defined here it seemed to make sense to move so all the `cfg` gates combine.

(the motivation of feature-flagging this is we can drop the `mc-account-keys` dependency completely when not required)

File: transaction/summary/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+mod data;
+mod error;
+mod report;
+mod verifier;
+
+#[cfg(feature = "mc-account-keys")]
+pub use data::{verify_tx_summary, TxOutSummaryUnblindingData, TxSummaryUnblindingData};
+pub use error::Error;
+pub use report::{TransactionEntity, TxSummaryUnblindingReport};
+pub use verifier::{TxSummaryStreamingVerifier, TxSummaryStreamingVerifierCtx};
```
Comment (ryankurte) [inline transaction/summary/src/lib.rs:17]: in ring signature we have the `internals` flag to gate access to streaming primitives, could repeat this here for `TxSummaryStreamingVerifierCtx`?

File: crypto/ring-signature/src/ring_signature/mlsag.rs [hunk 4]
```diff
@@ -270,7 +270,7 @@ mod mlsag_tests {
             };
 
             let real_index = rng.next_u64() as usize % (num_mixins + 1);
-            ring.insert(real_index, reduced_tx_out);
+            let _ = ring.insert(real_index, reduced_tx_out);
```
Comment (cbeck88) [inline crypto/ring-signature/src/ring_signature/mlsag.rs:273]: I'm a bit confused by this, these are heapless errors that we were ignoring before?

I think maybe it's better to panic rather than swallowing these errors?

File: transaction/summary/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,49 @@
+[package]
+name = "mc-transaction-summary"
+version = "4.0.2"
+authors = ["MobileCoin"]
+edition = "2021"
+readme = "README.md"
```
Comment (cbeck88) [inline transaction/summary/Cargo.toml:6]: it looks that we declared a README, but I don't see one in the diff? is that right? can we add one?

File: transaction/summary/src/lib.rs (hunk 3 above)
Comment (cbeck88) [reply transaction/summary/src/lib.rs:17]: I'm a bit skeptical of `internals` flags -- either a piece of API is public, or it isn't. Cynically, an internals flag is kind of like a weird middle ground to avoid making a decision. I think we should just say that, these types are public, and simply document that most users won't need them.

File: transaction/summary/src/verifier.rs [hunk 6]
```diff
@@ -117,7 +43,17 @@ pub fn verify_tx_summary(
...
+    // The report which we produce about what balance changes occur for what
+    // parties
+    report: TxSummaryUnblindingReport<RECORDS>,
+
+    // Context for streaming verifier
+    ctx: TxSummaryStreamingVerifierCtx,
+}
+
+/// Context for computing streaming transaction summaries
+pub struct TxSummaryStreamingVerifierCtx {
```
Comment (cbeck88) [inline transaction/summary/src/verifier.rs:56]: maybe this should also have a `MAX_RECORDS` parameter (and like phantom-data I guess?) to avoid mix and matching report sizes across the different api calls to this function?

not sure, maybe it's nicer the way you did it

File: transaction/summary/src/verifier.rs (hunk 1 above)
Comment (cbeck88) [reply transaction/summary/src/verifier.rs:46]: This is a bit surprising to me -- when I wrote the code, `verify_tx_summary` was only there for ease of testing

I need to read the PR more carefully still

File: crypto/dalek/Cargo.toml [hunk 7]
```diff
@@ -24,8 +25,8 @@ x25519-dalek = { version = "2.0.0-pre.2", default-features = false, features = [
 # if `mc-crypto-dalek/default` is selected we'll pull in the avx or u64 backends based on the target,
 # otherwise backend selection is up to the user
 
-[target.'cfg(any(target_feature = "avx2", target_feature = "avx"))'.dependencies]
+[target.'cfg(all(any(target_feature = "avx2", target_feature = "avx"), target_arch = "x86_64"))'.dependencies]
```
Comment (cbeck88) [inline crypto/dalek/Cargo.toml:28]: are you sure this is correct? I think not all x86_64 necessarily have avx2

File: transaction/summary/src/verifier.rs (hunk 1 above)
Comment (cbeck88) [reply transaction/summary/src/verifier.rs:46]: I don't understand, does it mean that you only use `verify_tx_summary` in the hardware wallet?

The `verify_tx_summary` requires all the `TxSummary` and `TxSummaryUnblindingData` to be available, so does that mean we don't stream it?

**Review by cbeck88 — CHANGES_REQUESTED** [review]
left some questions

File: transaction/summary/src/verifier.rs (hunk 1 above)
Comment (cbeck88) [reply transaction/summary/src/verifier.rs:46]: I think even if you aren't using it, it still makes sense to make the `TxSummaryStreamingVerifier` public in case someone else wants to actually take advantage of the streaming part.

Comment (ryankurte) [reply transaction/summary/src/verifier.rs:46]: yeah the naming is confusing atm (sorry)... `TxSummaryStreamingVerifierCtx` is the streaming verifier context, used by the hardware wallet, `TxSummaryStreamingVerifier` just combines that with the report for easier handling, which is then called in `verify_tx_summary`.

i think probably i should remove the `TxSummaryStreamingVerifier` impl because if you need that function it's already represented by the `TxSummaryStreamingVerifierCtx` object and it's not much more complex to call this more directly from `verify_tx_summary`

File: transaction/summary/src/lib.rs (hunk 3 above)
Comment (ryankurte) [reply transaction/summary/src/lib.rs:17]: i think the justification wrt. the ring signer was that you can misuse the streaming version in ways that the higher level API precludes, like misconstructing a ring or a hash is easier with the streaming version, but okay either way

File: crypto/ring-signature/src/ring_signature/mlsag.rs (hunk 4 above)
Comment (ryankurte) [reply crypto/ring-signature/src/ring_signature/mlsag.rs:273]: yeah updated clippy lint was giving me failures... i'll go back and make sure this is the case but these errors _should_ be unreachable because we fail on extending the vec to the required ring length earlier in the process. prefer a panic or a propagated error?

File: transaction/summary/src/verifier.rs (hunk 6 above)
Comment (ryankurte) [reply transaction/summary/src/verifier.rs:56]: ahh i also preferred this at the top level too, iirc got in a fight with clippy over trying to hold a const generic without actually _using_ it / binding the global N to the report argument N and couldn't convince `PhantomData` to hold it...

can revisit and see if there's a better option, alternatively we could abstract report into a trait so the verifier doesn't have visibility into the storage at all?

File: crypto/dalek/Cargo.toml (hunk 7 above)
Comment (ryankurte) [reply crypto/dalek/Cargo.toml:28]: yep unless i've flubbed this is `(avx2 || avx) && x86_64` because `avx` extensions are `x86_64` specific and some broken target detection somewhere was trying to invoke the avx2 feature when building for the ledger

Comment (cbeck88) [reply crypto/dalek/Cargo.toml:28]: oh i misread it, i thought this was `avx2 || avx || x86_64`

okay i get what you are doing

sounds good to me, thanks

File: transaction/summary/src/verifier.rs (hunk 6 above)
Comment (cbeck88) [reply transaction/summary/src/verifier.rs:56]: no strong preference, happy to dodge a compiler fight 😄

File: transaction/summary/src/lib.rs (hunk 3 above)
Comment (cbeck88) [reply transaction/summary/src/lib.rs:17]: okay, i see that.

Another drawback of actually having a `cargo` feature called `internals` is that it increases your testing surface area. Now you probably want to test that the code builds with the flag on and off.

In other projects, like prost for instance, the way they deal with internals is they use a `#[doc(hidden)]` attribute, so it's just not documented.

In serde, they just have a module called `serde::__private` and the internals are there.

Neither of these creates an additional cargo feature, so it dodges that particular problem

Comment (cbeck88) [reply transaction/summary/src/lib.rs:17]: My bias would be, if it's more possible to misconstruct a ring with a given API, let's just document that fact, and skip the "internals" feature flag

File: crypto/ring-signature/src/ring_signature/mlsag.rs (hunk 4 above)
Comment (cbeck88) [reply crypto/ring-signature/src/ring_signature/mlsag.rs:273]: if it's like, a numeric limit is exceeded, and we should know statically that it can't be reached, then a panic seems correct, since it's a logic error?

if possible, we could pair that with like a `static_assert`? (I haven't read all the surrounding code)

Comment (cbeck88) [reply crypto/ring-signature/src/ring_signature/mlsag.rs:273]: if you really think it's unreachable then panic seems good

File: transaction/summary/src/lib.rs (hunk 3 above)
Comment (jcape) [reply transaction/summary/src/lib.rs]: Sounds like a job for `"hazmat"`?

**Comment by cbeck88** [comment]
If you want to add flags like `internals` can you please update CI so that it tests that the code builds with the flag on and off?

**Comment by jcape** [comment]
Any reason to avoid calling this `hazmat` instead of `internals`, ala https://docs.rs/aes/latest/aes/hazmat/index.html

File: transaction/summary/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+#![no_std]
+
```
Comment (cbeck88) [inline transaction/summary/src/lib.rs:4]:
```suggestion
#![doc = include_str!("../README.md")]
#![deny(missing_docs)]

```

File: crypto/ring-signature/src/ring_signature/mlsag.rs (hunk 4 above)
Comment (ryankurte) [reply crypto/ring-signature/src/ring_signature/mlsag.rs]: dropped these changes for the moment

**Comment by ryankurte** [comment]
-gettin in fights with rebase, please standby-

File: crypto/ring-signature/src/ring_signature/mlsag.rs (hunk 4 above)
Comment (cbeck88) [reply crypto/ring-signature/src/ring_signature/mlsag.rs]: Let's make a github issue or something to track it?

File: crypto/digestible/Cargo.toml [hunk 9]
```diff
@@ -24,8 +24,8 @@ x25519-dalek = { version = "2.0.0-pre.2", default-features = false, optional = t
 [features]
 default = ["alloc", "derive", "dalek", "mc-crypto-dalek/default"]
 # Enables support for types in alloc crate
-alloc = ["mc-crypto-dalek/alloc"]
+alloc = []
 # Enables re-export of derive(Digestible) macro
 derive = ["mc-crypto-digestible-derive"]
 # Enables support for some crypto primitives in dalek crates
-dalek = ["mc-crypto-dalek", "curve25519-dalek", "ed25519-dalek", "x25519-dalek"]
+dalek = ["curve25519-dalek", "ed25519-dalek", "x25519-dalek"]
```
[the remark this answers is not in the document]
Comment (jcape) [reply crypto/digestible/Cargo.toml:31]: For future reference, I'd also suggest not mixing/matching multiple things in the same PR, e.g. fiddling with all this can/should be it's own PR, aside from moving tx computation to a separate crate. That way, the separation can continue, and the dalek mess can get cleaned up independently.

File: transaction/summary/src/data.rs [hunk 10]
```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use alloc::vec::Vec;
+
+use crate::TxSummaryStreamingVerifierCtx;
+
+use super::{Error, TxSummaryUnblindingReport};
```
Comment (jcape) [inline transaction/summary/src/data.rs:7]:
```suggestion
use crate::TxSummaryStreamingVerifierCtx;
use super::{Error, TxSummaryUnblindingReport};
use alloc::vec::Vec;```

File: transaction/summary/src/report.rs [hunk 11]
```diff
@@ -4,8 +4,10 @@
 //! about a transaction.
 
 use super::Error;
-use core::fmt::Display;
 use displaydoc::Display;
+
+use core::writeln;
+
```
Comment (jcape) [inline transaction/summary/src/report.rs:10]: This was fine as-is, please revert this change.

File: transaction/summary/src/verifier.rs [hunk 12]
```diff
@@ -1,3 +1,5 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
 // Copyright (c) 2018-2022 The MobileCoin Foundation
```
Comment (jcape) [inline transaction/summary/src/verifier.rs:3]: ?

**Review by jcape — CHANGES_REQUESTED** [review]

File: transaction/extra/src/unsigned_tx.rs [hunk 13]
```diff
@@ -11,7 +10,9 @@ use mc_transaction_core::{
     tx::{Tx, TxPrefix},
     FeeMap,
 };
+use mc_transaction_summary::{TxOutSummaryUnblindingData, TxSummaryUnblindingData};
 use mc_transaction_types::{Amount, BlockVersion, TokenId, TxSummary, UnmaskedAmount};
+
```
Comment (cbeck88) [inline transaction/extra/src/unsigned_tx.rs:15]: our style-guide says not to insert blank lines like this, we want cargo fmt to sort them

File: transaction/summary/src/verifier.rs [hunk 14]
```diff
@@ -155,7 +84,7 @@ impl TxSummaryStreamingVerifier {
     /// Returns:
     /// * A properly initialized TxSummaryStreamingVerifier
     pub fn new(
-        extended_message_digest: &[u8; 32],
+        extended_message_digest: &[u8],
```
Comment (cbeck88) [inline transaction/summary/src/verifier.rs:87]: is there a good reason for this change? i thought we reviewed this earlier and reverted it

Comment (ryankurte) [reply transaction/summary/src/verifier.rs:87]: ahhh we did?! i must have done _something_ wrong when restacking :-/

File: crypto/digestible/Cargo.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jcape) [reply crypto/digestible/Cargo.toml:31]: I mean, OK, that's a bug in `mc-crypto-dalek`, though it sounds like something that's worth fixing after we've unforked `curve25519-dalek` to 4.0.0-rc.1. Any reason not to just prioritize that work ahead of this, then fix this to handle it?

**Review by cbeck88 — APPROVED** [review]

**Comment by ryankurte** [comment]
-rebasing against master, seem to have pulled in some more changes than expected that i'm following up atm-

**Review by jcape — APPROVED** [review]
